File: examples/sidecar-js/decision-audit-reporter/test/boundary-tester.test.ts
import { describe, expect, it, vi } from 'vitest';

import { BoundaryTester } from '../src/boundary-tester.js';
import type { SidecarPlugin } from '../src/types.js';

describe('BoundaryTester', () => {
  it('evaluates decision with inputs via mocked plugin and records success results', async () => {
    const evaluateDecision = vi.fn(async (_modelId: string, input: Record<string, unknown>) => ({
      result: { tier: 'gold' },
      matchedRules: [{ rule_id: 'rule_2' }],
    }));

    const plugin: SidecarPlugin = {
      connect: async () => {},
      register: async () => {},
      onEvent: () => {},
      evaluateDecision,
      getFlowNodeInstance: async () => null,
      disconnect: async () => {},
    };

    const tester = new BoundaryTester(plugin);
    const results = await tester.testAll([
      {
        decisionRef: 'employee-benefits',
        boundaryInputs: [
          {
            testCase: 'gold_outstanding',
            input: {
              yearsOfService: 12,
              department: 'sales',
              performanceRating: 'outstanding',
              employeeType: 'full_time',
            },
          },
        ],
      },
    ]);

    const modelResults = results.get('employee-benefits');
    expect(modelResults).toHaveLength(1);
    expect(modelResults![0]).toEqual({
      testCase: 'gold_outstanding',
      input: {
        yearsOfService: 12,
        department: 'sales',
        performanceRating: 'outstanding',
        employeeType: 'full_time',
      },
      result: { tier: 'gold' },
      error: null,
    });
    expect(evaluateDecision).toHaveBeenCalledWith(
      'employee-benefits',
      modelResults![0].input,
    );
  });

  it('records error results when evaluation fails', async () => {
    const plugin: SidecarPlugin = {
      connect: async () => {},
      register: async () => {},
      onEvent: () => {},
      evaluateDecision: vi.fn(async () => {
        throw new Error('dmn_evaluation_error');
      }),
      getFlowNodeInstance: async () => null,
      disconnect: async () => {},
    };

    const tester = new BoundaryTester(plugin);
    const results = await tester.testAll([
      {
        decisionRef: 'employee-benefits',
        boundaryInputs: [
          { testCase: 'invalid_input', input: { yearsOfService: -1 } },
        ],
      },
    ]);

    const modelResults = results.get('employee-benefits');
    expect(modelResults![0].result).toBeNull();
    expect(modelResults![0].error).toBe('Error: dmn_evaluation_error');
  });
});

File: examples/plugins/business_rules/decision_audit_reporter/test/boundary_tester_test.exs
defmodule Examples.BusinessRules.DecisionAuditReporter.BoundaryTesterTest do
  use ExUnit.Case, async: true

  alias BfwEngine.EngineFacade
  alias Examples.BusinessRules.DecisionAuditReporter.BoundaryTester

  test "evaluates decision with inputs via facade and records success results" do
    {:ok, calls_agent} = Agent.start_link(fn -> [] end)

    facade = %EngineFacade{
      engine_id: "test-engine-id",
      engine_name: "test-engine-name",
      version: "0.0.0-test",
      decisions: %EngineFacade.Decisions{
        evaluate: fn decision_ref, input, _options ->
          Agent.update(calls_agent, fn calls -> [{decision_ref, input} | calls] end)
          {:ok, %{result: %{"tier" => "gold"}, matched_rules: [%{rule_id: "rule_2"}]}}
        end
      }
    }

    input = %{
      "yearsOfService" => 12,
      "department" => "sales",
      "performanceRating" => "outstanding",
      "employeeType" => "full_time"
    }

    results =
      BoundaryTester.test_all(facade, [
        %{
          decision_ref: "employee-benefits",
          boundary_inputs: [%{test_case: "gold_outstanding", input: input}]
        }
      ])

    model_results = results["employee-benefits"]
    assert length(model_results) == 1

    assert hd(model_results) == %{
             test_case: "gold_outstanding",
             input: input,
             result: %{"tier" => "gold"},
             error: nil
           }

    assert Agent.get(calls_agent, & &1) == [{"employee-benefits", input}]
  end

  test "records error results when evaluation fails" do
    facade = %EngineFacade{
      engine_id: "test-engine-id",
      engine_name: "test-engine-name",
      version: "0.0.0-test",
      decisions: %EngineFacade.Decisions{
        evaluate: fn _decision_ref, _input, _options ->
          {:error, :dmn_evaluation_error}
        end
      }
    }

    results =
      BoundaryTester.test_all(facade, [
        %{
          decision_ref: "employee-benefits",
          boundary_inputs: [
            %{test_case: "invalid_input", input: %{"yearsOfService" => -1}}
          ]
        }
      ])

    model_result = hd(results["employee-benefits"])
    assert model_result.result == nil
    assert model_result.error == ":dmn_evaluation_error"
  end
end
